File: src/main/java/fr/lewon/web/bot/util/HHRequestProcessor.java
package fr.lewon.web.bot.util;

import java.util.ArrayList;
import java.util.List;

import org.apache.http.Header;
import org.apache.http.HttpResponse;
import org.apache.http.message.BasicHeader;

import fr.lewon.bot.http.AbstractRequestProcessor;
import fr.lewon.bot.http.body.HttpBodyBuilder;
import fr.lewon.bot.http.body.urlencoded.FUEBuilder;
import fr.lewon.web.bot.entities.activities.Mission;
import fr.lewon.web.bot.entities.battle.BattleMob;
import fr.lewon.web.bot.entities.battle.BattlePlayer;
import fr.lewon.web.bot.entities.enums.Currency;
import fr.lewon.web.bot.entities.input.battle.ActionBattleMob;
import fr.lewon.web.bot.entities.input.battle.ActionBattlePlayer;
import fr.lewon.web.bot.entities.input.champions.ActionChampionsTeamDraft;
import fr.lewon.web.bot.entities.input.contest.ActionContestGiveReward;
import fr.lewon.web.bot.entities.input.girl.ActionGirlAllSalaries;
import fr.lewon.web.bot.entities.input.girl.ActionGirlSingleSalary;
import fr.lewon.web.bot.entities.input.leagues.ActionLeaguesGetOpponentInfo;
import fr.lewon.web.bot.entities.input.mission.ActionMissionClaimReward;
import fr.lewon.web.bot.entities.input.mission.ActionMissionGiveGift;
import fr.lewon.web.bot.entities.input.mission.ActionMissionStartMission;
import fr.lewon.web.bot.entities.input.others.PlayerInfos;
import fr.lewon.web.bot.entities.input.quest.ActionQuestNext;
import fr.lewon.web.bot.entities.input.shop.ActionBuyItem;
import fr.lewon.web.bot.entities.input.stat.ActionUpgradeStat;
import fr.lewon.web.bot.entities.input.teambattle.ActionTeamBattleChampion;
import fr.lewon.web.bot.entities.response.DraftResponse;
import fr.lewon.web.bot.entities.response.OpponentInfoResponse;
import fr.lewon.web.bot.entities.response.Response;
import fr.lewon.web.bot.entities.response.SalaryResponse;
import fr.lewon.web.bot.entities.response.TeamBattleResponse;
import fr.lewon.web.bot.entities.shop.Item;

public class HHRequestProcessor extends AbstractRequestProcessor {

	private static final String BASE_URL = "https://www.hentaiheroes.com/";

	private static final String PHOENIX_AJAX = "/phoenix-ajax.php";
	private static final String AJAX = "/ajax.php";

	private static final String HOME = "/home.html";
	private static final String SHOP = "/shop.html";
	private static final String HAREM = "/harem.html";
	private static final String BATTLE = "/battle.html";
	private static final String ARENA = "/arena.html";
	private static final String TOWER_OF_FAME = "/tower-of-fame.html";
	private static final String ACTIVITIES = "/activities.html";
	private static final String CHAMPIONS_MAP = "champions-map.html";
	private static final String MAP = "/map.html";
	private static final String WORLD = "/world";
	private static final String CHAMPIONS = "/champions";
	private static final String QUEST = "/quest";

	private static final String SLASH = "/";

	
	private HttpBodyBuilder bodyBuilder = new FUEBuilder();
	
	@Override
	protected List<Header> getNeededHeaders() {
		List<Header> neededHeaders = new ArrayList<>();
		neededHeaders.add(new BasicHeader("Accept", "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3"));
		neededHeaders.add(new BasicHeader("Accept-Language", "fr-FR,fr;q=0.9,en-GB;q=0.8,en;q=0.7,en-US;q=0.6"));
		neededHeaders.add(new BasicHeader("Cache-Control", "max-age=0"));
		neededHeaders.add(new BasicHeader("Connection", "keep-alive"));
		neededHeaders.add(new BasicHeader("Content-Type", "application/x-www-form-urlencoded; charset=UTF-8"));
		neededHeaders.add(new BasicHeader("Host", "www.hentaiheroes.com"));
		neededHeaders.add(new BasicHeader("Upgrade-Insecure-Requests", "1"));
		neededHeaders.add(new BasicHeader("Origin", "https://www.hentaiheroes.com"));
		neededHeaders.add(new BasicHeader("Referer", "https://www.hentaiheroes.com/harem/1"));
		neededHeaders.add(new BasicHeader("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/72.0.3626.121 Safari/537.36"));
		return neededHeaders;
	}

	public HttpResponse getSession(String login, String password) throws Exception {
		return processPostRequest(BASE_URL + PHOENIX_AJAX, 
				bodyBuilder.generateBody(new PlayerInfos(login, password)));
	}

	public String getHomeContent(HHSession session) throws Exception {
		return readAllPageContent(BASE_URL + HOME, session.getCookieHeader());
	}

	public String getHaremContent(HHSession session) throws Exception {
		return readAllPageContent(BASE_URL + HAREM, session.getCookieHeader());
	}

	public String getMapContent(HHSession session) throws Exception {
		return readAllPageContent(BASE_URL + MAP, session.getCookieHeader());
	}

	public String getActivitiesContent(HHSession session) throws Exception {
		return readAllPageContent(BASE_URL + ACTIVITIES, session.getCookieHeader());
	}
	
	public String getChampionsMapContent(HHSession session) throws Exception {
		return readAllPageContent(BASE_URL + CHAMPIONS_MAP, session.getCookieHeader());
	}
	
	public String getChampionPageContent(HHSession session, Integer championId) throws Exception {
		return readAllPageContent(BASE_URL + CHAMPIONS + SLASH + championId, session.getCookieHeader());
	}

	public String getTowerOfFameContent(HHSession session) throws Exception {
		return readAllPageContent(BASE_URL + TOWER_OF_FAME, session.getCookieHeader());
	}

	public String getArenaContent(HHSession session) throws Exception {
		return readAllPageContent(BASE_URL + ARENA, session.getCookieHeader());
	}

	public String getWorldContent(HHSession session, String idWorld) throws Exception {
		return readAllPageContent(BASE_URL + WORLD + SLASH + idWorld, session.getCookieHeader());
	}

	public String getBattleArenaContent(HHSession session, int idArena) throws Exception {
		return readAllPageContent(BASE_URL + BATTLE + "?id_arena=" + idArena, session.getCookieHeader());
	}

	public String getBattleTrollContent(HHSession session, String idTroll) throws Exception {
		return readAllPageContent(BASE_URL + BATTLE + "?id_troll=" + idTroll, session.getCookieHeader());
	}

	public String getLeagueBattleContent(HHSession session, String id) throws Exception {
		return readAllPageContent(BASE_URL + BATTLE + "?league_battle=1&id_member=" + id, session.getCookieHeader());
	}

	public String getQuestContent(HHSession session, Long questId) throws Exception {
		return readAllPageContent(BASE_URL + QUEST + SLASH + questId, session.getCookieHeader());
	}
	
	public String getShopContent(HHSession session) throws Exception {
		return readAllPageContent(BASE_URL + SHOP, session.getCookieHeader());
	}

	public OpponentInfoResponse getOpponentInfo(HHSession session, String opponentId) throws Exception {
		return processPostRequest(OpponentInfoResponse.class, BASE_URL + AJAX, 
				bodyBuilder.generateBody(new ActionLeaguesGetOpponentInfo(opponentId)), session.getCookieHeader()).getEntity();
	}

	public SalaryResponse getSalary(HHSession session, int which) throws Exception {
		return processPostRequest(SalaryResponse.class, BASE_URL + AJAX, 
				bodyBuilder.generateBody(new ActionGirlSingleSalary(which)), session.getCookieHeader()).getEntity();
	}

	public SalaryResponse getAllSalaries(HHSession session) throws Exception {
		return processPostRequest(SalaryResponse.class, BASE_URL + AJAX, 
				bodyBuilder.generateBody(new ActionGirlAllSalaries()), session.getCookieHeader()).getEntity();
	}
	
	public DraftResponse draftChampionFight(HHSession session, Integer idChampion, List<Integer> girlsToKeep) throws Exception {
		return processPostRequest(DraftResponse.class, BASE_URL + AJAX,
				bodyBuilder.generateBody(new ActionChampionsTeamDraft(idChampion, girlsToKeep)), session.getCookieHeader()).getEntity();
	}
	
	public TeamBattleResponse fightChampion(HHSession session, Currency currency, Integer championId, List<Integer> girls) throws Exception {
		return processPostRequest(TeamBattleResponse.class, BASE_URL + AJAX,
				bodyBuilder.generateBody(new ActionTeamBattleChampion(currency, championId, girls)), session.getCookieHeader()).getEntity();
	}

	public Response startMission(HHSession session, Mission mission) throws Exception {
		return processPostRequest(SalaryResponse.class, BASE_URL + AJAX, 
				bodyBuilder.generateBody(new ActionMissionStartMission(mission.getIdMission(), mission.getIdMemberMission())), session.getCookieHeader()).getEntity();
	}

	public Response getFinalMissionGift(HHSession session) throws Exception {
		return processPostRequest(SalaryResponse.class, BASE_URL + AJAX, 
				bodyBuilder.generateBody(new ActionMissionGiveGift()), session.getCookieHeader()).getEntity();
	}

	public Response claimReward(HHSession session, Mission mission) throws Exception {
		return processPostRequest(SalaryResponse.class, BASE_URL + AJAX, 
				bodyBuilder.generateBody(new ActionMissionClaimReward(mission.getIdMission(), mission.getIdMemberMission())), session.getCookieHeader()).getEntity();
	}

	public Response fightOpponentPlayer(HHSession session, BattlePlayer battlePlayer) throws Exception {
		return processPostRequest(Response.class, BASE_URL + AJAX, 
				bodyBuilder.generateBody(new ActionBattlePlayer(battlePlayer)), session.getCookieHeader()).getEntity();
	}

	public Response fightOpponentMob(HHSession session, BattleMob battleMob) throws Exception {
		return processPostRequest(Response.class, BASE_URL + AJAX, 
				bodyBuilder.generateBody(new ActionBattleMob(battleMob)), session.getCookieHeader()).getEntity();
	}

	public Response continueQuest(HHSession session, Long questId) throws Exception {
		return processPostRequest(Response.class, BASE_URL + AJAX, 
				bodyBuilder.generateBody(new ActionQuestNext(questId)), session.getCookieHeader()).getEntity();
	}

	public Response upgradeStat(HHSession session, int statToUpgrade) throws Exception {
		return processPostRequest(Response.class, BASE_URL + AJAX, 
				bodyBuilder.generateBody(new ActionUpgradeStat(statToUpgrade)), session.getCookieHeader()).getEntity();
	}

	public Response buyItem(Item item, HHSession session) throws Exception {
		return processPostRequest(Response.class, BASE_URL + AJAX,
				bodyBuilder.generateBody(new ActionBuyItem(item)), session.getCookieHeader()).getEntity();
	}

	public Response collectCompetitionRewards(Integer idContest, HHSession session) throws Exception {
		return processPostRequest(Response.class, BASE_URL + AJAX,
				bodyBuilder.generateBody(new ActionContestGiveReward(idContest)), session.getCookieHeader()).getEntity();
	}


}

File: src/main/kotlin/fr/lewon/bot/hh/rest/HHRequestProcessor.kt
package fr.lewon.bot.hh.rest

import fr.lewon.bot.hh.entities.activities.Mission
import fr.lewon.bot.hh.entities.battle.BattleMob
import fr.lewon.bot.hh.entities.battle.BattlePlayer
import fr.lewon.bot.hh.entities.battle.BattlePlayerSeason
import fr.lewon.bot.hh.entities.enums.Currency
import fr.lewon.bot.hh.entities.input.battle.ActionBattleMob
import fr.lewon.bot.hh.entities.input.battle.ActionBattlePlayer
import fr.lewon.bot.hh.entities.input.battle.ActionBattlePlayerSeason
import fr.lewon.bot.hh.entities.input.champions.ActionChampionsTeamDraft
import fr.lewon.bot.hh.entities.input.contest.ActionContestGiveReward
import fr.lewon.bot.hh.entities.input.girl.ActionGirlAllSalaries
import fr.lewon.bot.hh.entities.input.girl.ActionGirlSingleSalary
import fr.lewon.bot.hh.entities.input.leagues.ActionLeagueReward
import fr.lewon.bot.hh.entities.input.leagues.ActionLeaguesGetOpponentInfo
import fr.lewon.bot.hh.entities.input.mission.ActionMissionClaimReward
import fr.lewon.bot.hh.entities.input.mission.ActionMissionGiveGift
import fr.lewon.bot.hh.entities.input.mission.ActionMissionStartMission
import fr.lewon.bot.hh.entities.input.others.PlayerInfo
import fr.lewon.bot.hh.entities.input.powerplaces.ActionPlaceOfPowerCollect
import fr.lewon.bot.hh.entities.input.powerplaces.ActionPlaceOfPowerStart
import fr.lewon.bot.hh.entities.input.quest.ActionQuestNext
import fr.lewon.bot.hh.entities.input.rewards.ActionWeeklyReward
import fr.lewon.bot.hh.entities.input.shop.ActionBuyItem
import fr.lewon.bot.hh.entities.input.stat.ActionUpgradeStat
import fr.lewon.bot.hh.entities.input.teambattle.ActionTeamBattleChampion
import fr.lewon.bot.hh.entities.response.*
import fr.lewon.bot.hh.entities.shop.Item
import fr.lewon.bot.runner.session.helpers.FUEBodyBuilder
import org.springframework.util.LinkedMultiValueMap
import org.springframework.web.reactive.function.client.WebClient

class HHRequestProcessor {

    private val cookieStore = LinkedMultiValueMap<String, String>()

    @Throws(Exception::class)
    fun login(webClient: WebClient, login: String, password: String) {
        cookieStore.clear()
        postForBody<String>(webClient, PHOENIX_AJAX, PlayerInfo(login, password))
        if (!cookieStore.contains("stay_online") || !cookieStore.contains("HH_SESS_13")) {
            throw Exception("Unable to connect to Hentai Heroes, website may be down")
        }
        cookieStore["lang"] = "fr"
        cookieStore["_pk_id.2.6e07"] = "cda6c741be7d8090.1562523528.2.1562529541.1562528524."
        cookieStore["_pk_ses.2.6e07"] = "1"
        cookieStore["age_verification"] = "1"
    }

    @Throws(Exception::class)
    fun getHomeContent(webClient: WebClient): String {
        return readAllPage(webClient, HOME)
    }

    @Throws(Exception::class)
    fun getHaremContent(webClient: WebClient): String {
        return readAllPage(webClient, HAREM)
    }

    @Throws(Exception::class)
    fun getMapContent(webClient: WebClient): String {
        return readAllPage(webClient, MAP)
    }

    @Throws(Exception::class)
    fun getActivitiesContent(webClient: WebClient): String {
        return readAllPage(webClient, ACTIVITIES)
    }

    @Throws(Exception::class)
    fun getChampionsMapContent(webClient: WebClient): String {
        return readAllPage(webClient, CHAMPIONS_MAP)
    }

    @Throws(Exception::class)
    fun getChampionPageContent(webClient: WebClient, championId: Int): String {
        return readAllPage(webClient, CHAMPIONS + SLASH + championId)
    }

    @Throws(Exception::class)
    fun getTowerOfFameContent(webClient: WebClient): String {
        return readAllPage(webClient, TOWER_OF_FAME)
    }

    @Throws(Exception::class)
    fun getArenaContent(webClient: WebClient): String {
        return readAllPage(webClient, ARENA)
    }

    @Throws(Exception::class)
    fun getWorldContent(webClient: WebClient, idWorld: String): String {
        return readAllPage(webClient, WORLD + SLASH + idWorld)
    }

    @Throws(Exception::class)
    fun getBattleArenaContent(webClient: WebClient, idArena: Int): String {
        return readAllPage(webClient, "$BATTLE?id_arena=$idArena")
    }

    @Throws(Exception::class)
    fun getBattleTrollContent(webClient: WebClient, idTroll: String): String {
        return readAllPage(webClient, "$BATTLE?id_troll=$idTroll")
    }

    @Throws(Exception::class)
    fun getLeagueBattleContent(webClient: WebClient, id: String): String {
        return readAllPage(webClient, "$BATTLE?league_battle=1&id_member=$id")
    }

    @Throws(Exception::class)
    fun getBattleSeasonArenaContent(webClient: WebClient, idSeasonArena: String): String {
        return readAllPage(webClient, "$BATTLE?id_season_arena=$idSeasonArena")
    }

    @Throws(Exception::class)
    fun getQuestContent(webClient: WebClient, questId: Long): String {
        return readAllPage(webClient, QUEST + SLASH + questId)
    }

    @Throws(Exception::class)
    fun getShopContent(webClient: WebClient): String {
        return readAllPage(webClient, SHOP)
    }

    @Throws(Exception::class)
    fun getSeasonContent(webClient: WebClient): String {
        return readAllPage(webClient, SEASON)
    }

    @Throws(Exception::class)
    fun getSeasonArenaContent(webClient: WebClient): String {
        return readAllPage(webClient, SEASON_ARENA)
    }

    @Throws(Exception::class)
    fun getOpponentInfo(webClient: WebClient, opponentId: String): OpponentInfoResponse? {
        return postForBody(webClient, AJAX, ActionLeaguesGetOpponentInfo(opponentId))
    }

    @Throws(Exception::class)
    fun getSalary(webClient: WebClient, which: Int): SalaryResponse? {
        return postForBody(webClient, AJAX, ActionGirlSingleSalary(which))
    }

    @Throws(Exception::class)
    fun getAllSalaries(webClient: WebClient): SalaryResponse? {
        return postForBody(webClient, AJAX, ActionGirlAllSalaries())
    }

    @Throws(Exception::class)
    fun draftChampionFight(webClient: WebClient, idChampion: Int, girlsToKeep: List<Int>): DraftResponse? {
        return postForBody(webClient, AJAX, ActionChampionsTeamDraft(idChampion, girlsToKeep))
    }

    @Throws(Exception::class)
    fun fightChampion(webClient: WebClient, currency: Currency, championId: Int, girls: List<Int>): TeamBattleResponse? {
        return postForBody(webClient, AJAX, ActionTeamBattleChampion(currency, championId, girls))
    }

    @Throws(Exception::class)
    fun startMission(webClient: WebClient, mission: Mission): Response? {
        return postForBody(webClient, AJAX, ActionMissionStartMission(mission))
    }

    @Throws(Exception::class)
    fun startPlaceOfPower(webClient: WebClient, placeOfPowerId: Int, selectedGirls: List<Int>): Response? {
        return postForBody(webClient, AJAX, ActionPlaceOfPowerStart(placeOfPowerId, selectedGirls))
    }

    @Throws(Exception::class)
    fun collectPlaceOfPower(webClient: WebClient, placeOfPowerId: Int): Response? {
        return postForBody(webClient, AJAX, ActionPlaceOfPowerCollect(placeOfPowerId))
    }

    @Throws(Exception::class)
    fun getFinalMissionGift(webClient: WebClient): Response? {
        return postForBody(webClient, AJAX, ActionMissionGiveGift())
    }

    @Throws(Exception::class)
    fun claimReward(webClient: WebClient, mission: Mission): Response? {
        return postForBody(webClient, AJAX, ActionMissionClaimReward(mission))
    }

    @Throws(Exception::class)
    fun fightOpponentPlayer(webClient: WebClient, battlePlayer: BattlePlayer): Response? {
        return postForBody(webClient, AJAX, ActionBattlePlayer(battlePlayer))
    }

    @Throws(Exception::class)
    fun fightOpponentMob(webClient: WebClient, battleMob: BattleMob): Response? {
        return postForBody(webClient, AJAX, ActionBattleMob(battleMob))
    }

    @Throws(Exception::class)
    fun fightOpponentSeason(webClient: WebClient, battlePlayerSeason: BattlePlayerSeason): Response? {
        return postForBody(webClient, AJAX, ActionBattlePlayerSeason(battlePlayerSeason))
    }

    @Throws(Exception::class)
    fun continueQuest(webClient: WebClient, questId: Long): Response? {
        return postForBody(webClient, AJAX, ActionQuestNext(questId))
    }

    @Throws(Exception::class)
    fun upgradeStat(webClient: WebClient, statToUpgrade: Int): Response? {
        return postForBody(webClient, AJAX, ActionUpgradeStat(statToUpgrade))
    }

    @Throws(Exception::class)
    fun buyItem(webClient: WebClient, item: Item): Response? {
        return postForBody(webClient, AJAX, ActionBuyItem(item))
    }

    @Throws(Exception::class)
    fun collectCompetitionRewards(webClient: WebClient, idContest: Int): Response? {
        return postForBody(webClient, AJAX, ActionContestGiveReward(idContest))
    }

    @Throws(Exception::class)
    fun claimWeeklyRewards(webClient: WebClient): Response? {
        return postForBody(webClient, AJAX, ActionWeeklyReward())
    }

    @Throws(Exception::class)
    fun claimLeagueRewards(webClient: WebClient): Response? {
        return postForBody(webClient, AJAX, ActionLeagueReward())
    }

    @Synchronized
    private fun readAllPage(webClient: WebClient, uri: String): String {
        val response = webClient.get()
                .uri(uri)
                .header("Cookie", cookieStore.entries.joinToString("; ") { (k, v) -> "$k=${v[0]}" })
                .exchange()
                .block()
        response?.cookies()?.forEach { (n, u) -> cookieStore[n] = u[0].value }
        return response?.bodyToMono(String::class.java)?.block() ?: ""
    }

    @Synchronized
    private fun <T> postForBody(webClient: WebClient, uri: String, body: Any, responseType: Class<T>): T? {
        val response = webClient.post()
                .uri(uri)
                .header("Cookie", cookieStore.entries.joinToString("; ") { (k, v) -> "$k=${v[0]}" })
                .bodyValue(FUEBodyBuilder().generateBody(body))
                .exchange().block()
        response?.cookies()?.forEach { (n, u) -> cookieStore[n] = u[0].value }
        return response?.bodyToMono(responseType)?.block()
    }

    private inline fun <reified T> postForBody(webClient: WebClient, uri: String, body: Any): T? {
        return postForBody(webClient, uri, body, T::class.java)
    }

    companion object {
        private const val PHOENIX_AJAX = "/phoenix-ajax.php"
        private const val AJAX = "/ajax.php"
        private const val HOME = "/home.html"
        private const val SHOP = "/shop.html"
        private const val HAREM = "/harem.html"
        private const val BATTLE = "/battle.html"
        private const val ARENA = "/arena.html"
        private const val TOWER_OF_FAME = "/tower-of-fame.html"
        private const val ACTIVITIES = "/activities.html"
        private const val CHAMPIONS_MAP = "champions-map.html"
        private const val MAP = "/map.html"
        private const val SEASON = "/season.html"
        private const val SEASON_ARENA = "/season-arena.html"
        private const val WORLD = "/world"
        private const val CHAMPIONS = "/champions"
        private const val QUEST = "/quest"
        private const val SLASH = "/"
    }
}